File: swf-parser.ts/src/lib/parsers/display.ts
import { Incident } from "incident";
import { Float32, Uint16, Uint32, Uint4, Uint5, Uint8, UintSize } from "semantic-types";
import {
  avm1,
  BlendMode,
  ClipActions,
  ClipEventFlags,
  ColorStop,
  Filter,
  filters,
  FilterType,
  Sfixed16P16,
  Sfixed8P8,
  StraightSRgba8,
} from "swf-tree";
import { ByteStream } from "../stream";
import { parseActionBlock } from "./avm1";
import { parseStraightSRgba8 } from "./basic-data-types";

export function parseBlendMode(byteStream: ByteStream): BlendMode {
  switch (byteStream.readUint8()) {
    case 0:
    case 1:
      return BlendMode.Normal;
    case 2:
      return BlendMode.Layer;
    case 3:
      return BlendMode.Multiply;
    case 4:
      return BlendMode.Screen;
    case 5:
      return BlendMode.Lighten;
    case 6:
      return BlendMode.Darken;
    case 7:
      return BlendMode.Difference;
    case 8:
      return BlendMode.Add;
    case 9:
      return BlendMode.Subtract;
    case 10:
      return BlendMode.Invert;
    case 11:
      return BlendMode.Alpha;
    case 12:
      return BlendMode.Erase;
    case 13:
      return BlendMode.Overlay;
    case 14:
      return BlendMode.Hardlight;
    default:
      throw new Incident("UnreachableCode");
  }
}

export function parseClipActionsString(byteStream: ByteStream, extendedEvents: boolean): ClipActions[] {
  byteStream.skip(2); // Reserved
  byteStream.skip(4); // All events (union of the events)
  const result: ClipActions[] = [];
  while (true) {
    const savedPos: UintSize = byteStream.bytePos;
    const peek: Uint32 = extendedEvents ? byteStream.readUint32BE() : byteStream.readUint32BE();
    if (peek === 0) {
      break;
    } else {
      byteStream.bytePos = savedPos;
    }
    result.push(parseClipActions(byteStream, extendedEvents));
  }

  return result;
}

export function parseClipEventFlags(byteStream: ByteStream, extendedEvents: boolean): ClipEventFlags {
  const flags: Uint32 = extendedEvents ? byteStream.readFloat32LE() : byteStream.readUint16LE();

  const load: boolean = (flags & (1 << 0)) !== 0;
  const enterFrame: boolean = (flags & (1 << 1)) !== 0;
  const unload: boolean = (flags & (1 << 2)) !== 0;
  const mouseMove: boolean = (flags & (1 << 3)) !== 0;
  const mouseDown: boolean = (flags & (1 << 4)) !== 0;
  const mouseUp: boolean = (flags & (1 << 5)) !== 0;
  const keyDown: boolean = (flags & (1 << 6)) !== 0;
  const keyUp: boolean = (flags & (1 << 7)) !== 0;
  const data: boolean = (flags & (1 << 8)) !== 0;
  const initialize: boolean = (flags & (1 << 9)) !== 0;
  const press: boolean = (flags & (1 << 10)) !== 0;
  const release: boolean = (flags & (1 << 11)) !== 0;
  const releaseOutside: boolean = (flags & (1 << 12)) !== 0;
  const rollOver: boolean = (flags & (1 << 13)) !== 0;
  const rollOut: boolean = (flags & (1 << 14)) !== 0;
  const dragOver: boolean = (flags & (1 << 15)) !== 0;
  const dragOut: boolean = (flags & (1 << 16)) !== 0;
  const keyPress: boolean = (flags & (1 << 17)) !== 0;
  const construct: boolean = (flags & (1 << 18)) !== 0;

  return {
    load,
    enterFrame,
    unload,
    mouseMove,
    mouseDown,
    mouseUp,
    keyDown,
    keyUp,
    data,
    initialize,
    press,
    release,
    releaseOutside,
    rollOver,
    rollOut,
    dragOver,
    dragOut,
    keyPress,
    construct,
  };
}

export function parseClipActions(byteStream: ByteStream, extendedEvents: boolean): ClipActions {
  const events: ClipEventFlags = parseClipEventFlags(byteStream, extendedEvents);
  let actionsSize: UintSize = byteStream.readUint32LE();
  let keyCode: Uint8 | undefined = undefined;
  if (events.keyPress) {
    keyCode = byteStream.readUint8();
    actionsSize -= 1;
  }
  const actions: avm1.Action[] = parseActionBlock(byteStream.take(actionsSize));
  return {events, keyCode, actions};
}

export function parseFilterList(byteStream: ByteStream): Filter[] {
  const filterCount: UintSize = byteStream.readUint8();
  const result: Filter[] = [];
  for (let i: number = 0; i < filterCount; i++) {
    result.push(parseFilter(byteStream));
  }
  return result;
}

export function parseFilter(byteStream: ByteStream): Filter {
  switch (byteStream.readUint8()) {
    case 0:
      return parseDropShadowFilter(byteStream);
    case 1:
      return parseBlurFilter(byteStream);
    case 2:
      return parseGlowFilter(byteStream);
    case 3:
      return parseBevelFilter(byteStream);
    case 4:
      return parseGradientGlowFilter(byteStream);
    case 5:
      return parseConvolutionFilter(byteStream);
    case 6:
      return parseColorMatrixFilter(byteStream);
    case 7:
      return parseGradientBevelFilter(byteStream);
    default:
      throw new Incident("UnreachableCode");
  }
}

export function parseBevelFilter(byteStream: ByteStream): filters.Bevel {
  const shadowColor: StraightSRgba8 = parseStraightSRgba8(byteStream);
  const highlightColor: StraightSRgba8 = parseStraightSRgba8(byteStream);
  const blurX: Sfixed16P16 = byteStream.readFixed16P16LE();
  const blurY: Sfixed16P16 = byteStream.readFixed16P16LE();
  const angle: Sfixed16P16 = byteStream.readFixed16P16LE();
  const distance: Sfixed16P16 = byteStream.readFixed16P16LE();
  const strength: Sfixed8P8 = byteStream.readFixed8P8LE();
  const flags: Uint8 = byteStream.readUint8();
  const inner: boolean = (flags & (1 << 7)) !== 0;
  const knockout: boolean = (flags & (1 << 6)) !== 0;
  const compositeSource: boolean = (flags & (1 << 5)) !== 0;
  const onTop: boolean = (flags & (1 << 4)) !== 0;
  const passes: Uint4 = <Uint4> (flags & ((1 << 4) - 1));
  return {
    filter: FilterType.Bevel,
    shadowColor,
    highlightColor,
    blurX,
    blurY,
    angle,
    distance,
    strength,
    inner,
    knockout,
    compositeSource,
    onTop,
    passes,
  };
}

export function parseBlurFilter(byteStream: ByteStream): filters.Blur {
  const blurX: Sfixed16P16 = byteStream.readFixed16P16LE();
  const blurY: Sfixed16P16 = byteStream.readFixed16P16LE();
  const flags: Uint8 = byteStream.readUint8();
  const passes: Uint5 = <Uint5> ((flags >> 3) & 0x1f);
  return {
    filter: FilterType.Blur,
    blurX,
    blurY,
    passes,
  };
}

export function parseColorMatrixFilter(byteStream: ByteStream): filters.ColorMatrix {
  const matrix: Float32[] = [];
  for (let i: number = 0; i < 20; i++) {
    matrix.push(byteStream.readFloat32LE());
  }
  return {
    filter: FilterType.ColorMatrix,
    matrix,
  };
}

export function parseConvolutionFilter(byteStream: ByteStream): filters.Convolution {
  const matrixWidth: UintSize = byteStream.readUint8();
  const matrixHeight: UintSize = byteStream.readUint8();
  const divisor: Float32 = byteStream.readFloat32LE();
  const bias: Float32 = byteStream.readFloat32LE();
  const matrix: Float32[] = [];
  for (let i: number = 0; i < matrixWidth * matrixHeight; i++) {
    matrix.push(byteStream.readFloat32LE());
  }
  const defaultColor: StraightSRgba8 = parseStraightSRgba8(byteStream);
  const flags: Uint8 = byteStream.readUint8();
  const clamp: boolean = (flags & (1 << 1)) !== 0;
  const preserveAlpha: boolean = (flags & (1 << 0)) !== 0;
  return {
    filter: FilterType.Convolution,
    matrixWidth,
    matrixHeight,
    divisor,
    bias,
    matrix,
    defaultColor,
    clamp,
    preserveAlpha,
  };
}

export function parseDropShadowFilter(byteStream: ByteStream): filters.DropShadow {
  const color: StraightSRgba8 = parseStraightSRgba8(byteStream);
  const blurX: Sfixed16P16 = byteStream.readFixed16P16LE();
  const blurY: Sfixed16P16 = byteStream.readFixed16P16LE();
  const angle: Sfixed16P16 = byteStream.readFixed16P16LE();
  const distance: Sfixed16P16 = byteStream.readFixed16P16LE();
  const strength: Sfixed8P8 = byteStream.readFixed8P8LE();
  const flags: Uint8 = byteStream.readUint8();
  const inner: boolean = (flags & (1 << 7)) !== 0;
  const knockout: boolean = (flags & (1 << 6)) !== 0;
  const compositeSource: boolean = (flags & (1 << 5)) !== 0;
  const passes: Uint5 = flags & ((1 << 5) - 1);
  return {
    filter: FilterType.DropShadow,
    color,
    blurX,
    blurY,
    angle,
    distance,
    strength,
    inner,
    knockout,
    compositeSource,
    passes,
  };
}

export function parseGlowFilter(byteStream: ByteStream): filters.Glow {
  const color: StraightSRgba8 = parseStraightSRgba8(byteStream);
  const blurX: Sfixed16P16 = byteStream.readFixed16P16LE();
  const blurY: Sfixed16P16 = byteStream.readFixed16P16LE();
  const strength: Sfixed8P8 = byteStream.readFixed8P8LE();
  const flags: Uint8 = byteStream.readUint8();
  const inner: boolean = (flags & (1 << 7)) !== 0;
  const knockout: boolean = (flags & (1 << 6)) !== 0;
  const compositeSource: boolean = (flags & (1 << 5)) !== 0;
  const passes: Uint5 = flags & ((1 << 5) - 1);
  return {
    filter: FilterType.Glow,
    color,
    blurX,
    blurY,
    strength,
    inner,
    knockout,
    compositeSource,
    passes,
  };
}

export function parseGradientBevelFilter(byteStream: ByteStream): filters.GradientBevel {
  const colorCount: UintSize = byteStream.readUint8();
  const gradient: ColorStop[] = [];
  for (let i: number = 0; i < colorCount; i++) {
    gradient.push({ratio: 0, color: parseStraightSRgba8(byteStream)});
  }
  for (let i: number = 0; i < colorCount; i++) {
    gradient[i].ratio = byteStream.readUint8();
  }
  const blurX: Sfixed16P16 = byteStream.readFixed16P16LE();
  const blurY: Sfixed16P16 = byteStream.readFixed16P16LE();
  const angle: Sfixed16P16 = byteStream.readFixed16P16LE();
  const distance: Sfixed16P16 = byteStream.readFixed16P16LE();
  const strength: Sfixed8P8 = byteStream.readFixed8P8LE();
  const flags: Uint8 = byteStream.readUint8();
  const inner: boolean = (flags & (1 << 7)) !== 0;
  const knockout: boolean = (flags & (1 << 6)) !== 0;
  const compositeSource: boolean = (flags & (1 << 5)) !== 0;
  const onTop: boolean = (flags & (1 << 4)) !== 0;
  const passes: Uint4 = <Uint4> (flags & ((1 << 4) - 1));
  return {
    filter: FilterType.GradientBevel,
    gradient,
    blurX,
    blurY,
    angle,
    distance,
    strength,
    inner,
    knockout,
    compositeSource,
    onTop,
    passes,
  };
}

export function parseGradientGlowFilter(byteStream: ByteStream): filters.GradientGlow {
  const colorCount: UintSize = byteStream.readUint8();
  const gradient: ColorStop[] = [];
  for (let i: number = 0; i < colorCount; i++) {
    gradient.push({ratio: 0, color: parseStraightSRgba8(byteStream)});
  }
  for (let i: number = 0; i < colorCount; i++) {
    gradient[i].ratio = byteStream.readUint8();
  }
  const blurX: Sfixed16P16 = byteStream.readFixed16P16LE();
  const blurY: Sfixed16P16 = byteStream.readFixed16P16LE();
  const angle: Sfixed16P16 = byteStream.readFixed16P16LE();
  const distance: Sfixed16P16 = byteStream.readFixed16P16LE();
  const strength: Sfixed8P8 = byteStream.readFixed8P8LE();
  const flags: Uint8 = byteStream.readUint8();
  const inner: boolean = (flags & (1 << 7)) !== 0;
  const knockout: boolean = (flags & (1 << 6)) !== 0;
  const compositeSource: boolean = (flags & (1 << 5)) !== 0;
  const onTop: boolean = (flags & (1 << 4)) !== 0;
  const passes: Uint4 = <Uint4> (flags & ((1 << 4) - 1));
  return {
    filter: FilterType.GradientGlow,
    gradient,
    blurX,
    blurY,
    angle,
    distance,
    strength,
    inner,
    knockout,
    compositeSource,
    onTop,
    passes,
  };
}

File: rs/src/complete/display.rs
use crate::complete::base::skip;
use crate::streaming::basic_data_types::{parse_le_fixed16_p16, parse_le_fixed8_p8, parse_straight_s_rgba8};
use nom::number::complete::{
  le_f32 as parse_le_f32, le_u16 as parse_le_u16, le_u32 as parse_le_u32, le_u8 as parse_u8,
};
use nom::IResult as NomResult;
use swf_types as swf;

#[allow(unused_variables)]
pub fn parse_blend_mode(input: &[u8]) -> NomResult<&[u8], swf::BlendMode> {
  let (input, code) = parse_u8(input)?;
  let blend_mode: swf::BlendMode = match code {
    0 => swf::BlendMode::Normal,
    1 => swf::BlendMode::Normal,
    2 => swf::BlendMode::Layer,
    3 => swf::BlendMode::Multiply,
    4 => swf::BlendMode::Screen,
    5 => swf::BlendMode::Lighten,
    6 => swf::BlendMode::Darken,
    7 => swf::BlendMode::Difference,
    8 => swf::BlendMode::Add,
    9 => swf::BlendMode::Subtract,
    10 => swf::BlendMode::Invert,
    11 => swf::BlendMode::Alpha,
    12 => swf::BlendMode::Erase,
    13 => swf::BlendMode::Overlay,
    14 => swf::BlendMode::Hardlight,
    _ => return Err(nom::Err::Error(nom::error::Error::new(input, nom::error::ErrorKind::Switch))),
  };
  Ok((input, blend_mode))
}

pub fn parse_clip_actions_string(input: &[u8], extended_events: bool) -> NomResult<&[u8], Vec<swf::ClipAction>> {
  use nom::combinator::map;

  let (input, _) = skip(2usize)(input)?; // Skip `reserved`
  let (input, _) = skip(if extended_events { 4usize } else { 2usize })(input)?; // Skip `all_events`

  let mut result: Vec<swf::ClipAction> = Vec::new();
  let mut current_input = input;

  loop {
    let head = if extended_events {
      parse_le_u32(current_input)
    } else {
      map(parse_le_u16, u32::from)(current_input)
    };

    match head {
      Ok((next_input, event_flags)) => {
        if event_flags == 0 {
          current_input = next_input;
          break;
        }
      }
      Err(e) => return Err(e),
    };

    match parse_clip_actions(current_input, extended_events) {
      Ok((next_input, clip_actions)) => {
        result.push(clip_actions);
        current_input = next_input;
      }
      Err(e) => return Err(e),
    };
  }

  Ok((current_input, result))
}

#[allow(unused_variables)]
pub fn parse_clip_event_flags(input: &[u8], extended_events: bool) -> NomResult<&[u8], swf::ClipEventFlags> {
  use nom::combinator::map;

  let (input, flags) = if extended_events {
    parse_le_u32(input)?
  } else {
    map(parse_le_u16, u32::from)(input)?
  };

  #[allow(clippy::identity_op)]
  Ok((
    input,
    swf::ClipEventFlags {
      load: (flags & (1 << 0)) != 0,
      enter_frame: (flags & (1 << 1)) != 0,
      unload: (flags & (1 << 2)) != 0,
      mouse_move: (flags & (1 << 3)) != 0,
      mouse_down: (flags & (1 << 4)) != 0,
      mouse_up: (flags & (1 << 5)) != 0,
      key_down: (flags & (1 << 6)) != 0,
      key_up: (flags & (1 << 7)) != 0,
      data: (flags & (1 << 8)) != 0,
      initialize: (flags & (1 << 9)) != 0,
      press: (flags & (1 << 10)) != 0,
      release: (flags & (1 << 11)) != 0,
      release_outside: (flags & (1 << 12)) != 0,
      roll_over: (flags & (1 << 13)) != 0,
      roll_out: (flags & (1 << 14)) != 0,
      drag_over: (flags & (1 << 15)) != 0,
      drag_out: (flags & (1 << 16)) != 0,
      key_press: (flags & (1 << 17)) != 0,
      construct: (flags & (1 << 18)) != 0,
    },
  ))
}

pub fn parse_clip_actions(input: &[u8], extended_events: bool) -> NomResult<&[u8], swf::ClipAction> {
  use nom::combinator::map;

  let (input, events) = parse_clip_event_flags(input, extended_events)?;
  let (input, actions_size) = map(parse_le_u32, |x| x as usize)(input)?;
  let (input, (actions_size, key_code)) = if events.key_press {
    let (input, key_code) = parse_u8(input)?;
    (input, (actions_size.saturating_sub(1), Some(key_code)))
  } else {
    (input, (actions_size, None))
  };
  let (input, actions) = nom::bytes::streaming::take(actions_size)(input)?;

  Ok((
    input,
    swf::ClipAction {
      events,
      key_code,
      actions: actions.to_vec(),
    },
  ))
}

pub fn parse_filter_list(input: &[u8]) -> NomResult<&[u8], Vec<swf::Filter>> {
  use nom::multi::count;
  let (input, filter_count) = parse_u8(input)?;
  count(parse_filter, usize::from(filter_count))(input)
}

#[allow(unused_variables)]
pub fn parse_filter(input: &[u8]) -> NomResult<&[u8], swf::Filter> {
  use nom::combinator::map;
  let (input, code) = parse_u8(input)?;
  match code {
    0 => map(parse_drop_shadow_filter, swf::Filter::DropShadow)(input),
    1 => map(parse_blur_filter, swf::Filter::Blur)(input),
    2 => map(parse_glow_filter, swf::Filter::Glow)(input),
    3 => map(parse_bevel_filter, swf::Filter::Bevel)(input),
    4 => map(parse_gradient_glow_filter, swf::Filter::GradientGlow)(input),
    5 => map(parse_convolution_filter, swf::Filter::Convolution)(input),
    6 => map(parse_color_matrix_filter, swf::Filter::ColorMatrix)(input),
    7 => map(parse_gradient_bevel_filter, swf::Filter::GradientBevel)(input),
    _ => Err(nom::Err::Error(nom::error::Error::new(input, nom::error::ErrorKind::Switch))),
  }
}

pub fn parse_bevel_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::Bevel> {
  let (input, shadow_color) = parse_straight_s_rgba8(input)?;
  let (input, highlight_color) = parse_straight_s_rgba8(input)?;
  let (input, blur_x) = parse_le_fixed16_p16(input)?;
  let (input, blur_y) = parse_le_fixed16_p16(input)?;
  let (input, angle) = parse_le_fixed16_p16(input)?;
  let (input, distance) = parse_le_fixed16_p16(input)?;
  let (input, strength) = parse_le_fixed8_p8(input)?;
  let (input, flags) = parse_u8(input)?;
  let passes = flags & 0b1111;
  let on_top = (flags & (1 << 4)) != 0;
  let composite_source = (flags & (1 << 5)) != 0;
  let knockout = (flags & (1 << 6)) != 0;
  let inner = (flags & (1 << 7)) != 0;

  Ok((
    input,
    swf::filters::Bevel {
      shadow_color,
      highlight_color,
      blur_x,
      blur_y,
      angle,
      distance,
      strength,
      inner,
      knockout,
      composite_source,
      on_top,
      passes,
    },
  ))
}

pub fn parse_blur_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::Blur> {
  let (input, blur_x) = parse_le_fixed16_p16(input)?;
  let (input, blur_y) = parse_le_fixed16_p16(input)?;
  let (input, flags) = parse_u8(input)?;
  // Skip bits [0, 2]
  let passes = flags >> 3;

  Ok((input, swf::filters::Blur { blur_x, blur_y, passes }))
}

pub fn parse_color_matrix_filter(mut input: &[u8]) -> NomResult<&[u8], swf::filters::ColorMatrix> {
  let mut matrix: [f32; 20] = [0f32; 20];
  for matrix_value in &mut matrix {
    let (next_input, value) = parse_le_f32(input)?;
    input = next_input;
    *matrix_value = value;
  }
  Ok((input, swf::filters::ColorMatrix { matrix }))
}

pub fn parse_convolution_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::Convolution> {
  use nom::combinator::map;
  use nom::multi::count;

  let (input, matrix_width) = map(parse_u8, usize::from)(input)?;
  let (input, matrix_height) = map(parse_u8, usize::from)(input)?;
  let (input, divisor) = parse_le_f32(input)?;
  let (input, bias) = parse_le_f32(input)?;
  let (input, matrix) = count(parse_le_f32, matrix_width * matrix_height)(input)?;
  let (input, default_color) = parse_straight_s_rgba8(input)?;
  let (input, flags) = parse_u8(input)?;
  #[allow(clippy::identity_op)]
  let preserve_alpha = (flags & (1 << 0)) != 0;
  let clamp = (flags & (1 << 1)) != 0;
  // Skip bits [2, 7]

  Ok((
    input,
    swf::filters::Convolution {
      matrix_width,
      matrix_height,
      divisor,
      bias,
      matrix,
      default_color,
      clamp,
      preserve_alpha,
    },
  ))
}

pub fn parse_drop_shadow_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::DropShadow> {
  let (input, color) = parse_straight_s_rgba8(input)?;
  let (input, blur_x) = parse_le_fixed16_p16(input)?;
  let (input, blur_y) = parse_le_fixed16_p16(input)?;
  let (input, angle) = parse_le_fixed16_p16(input)?;
  let (input, distance) = parse_le_fixed16_p16(input)?;
  let (input, strength) = parse_le_fixed8_p8(input)?;
  let (input, flags) = parse_u8(input)?;
  let passes = flags & ((1 << 5) - 1);
  let composite_source = (flags & (1 << 5)) != 0;
  let knockout = (flags & (1 << 6)) != 0;
  let inner = (flags & (1 << 7)) != 0;

  Ok((
    input,
    swf::filters::DropShadow {
      color,
      blur_x,
      blur_y,
      angle,
      distance,
      strength,
      inner,
      knockout,
      composite_source,
      passes,
    },
  ))
}

pub fn parse_glow_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::Glow> {
  let (input, color) = parse_straight_s_rgba8(input)?;
  let (input, blur_x) = parse_le_fixed16_p16(input)?;
  let (input, blur_y) = parse_le_fixed16_p16(input)?;
  let (input, strength) = parse_le_fixed8_p8(input)?;
  let (input, flags) = parse_u8(input)?;
  let passes = flags & ((1 << 5) - 1);
  let composite_source = (flags & (1 << 5)) != 0;
  let knockout = (flags & (1 << 6)) != 0;
  let inner = (flags & (1 << 7)) != 0;

  Ok((
    input,
    swf::filters::Glow {
      color,
      blur_x,
      blur_y,
      strength,
      inner,
      knockout,
      composite_source,
      passes,
    },
  ))
}

fn parse_filter_gradient(input: &[u8], color_count: usize) -> NomResult<&[u8], Vec<swf::ColorStop>> {
  let mut result: Vec<swf::ColorStop> = Vec::with_capacity(color_count);
  let mut current_input = input;

  for _ in 0..color_count {
    match parse_straight_s_rgba8(current_input) {
      Ok((next_input, color)) => {
        result.push(swf::ColorStop { ratio: 0, color });
        current_input = next_input;
      }
      Err(e) => return Err(e),
    };
  }
  for mut color_stop in &mut result {
    match parse_u8(current_input) {
      Ok((next_input, ratio)) => {
        color_stop.ratio = ratio;
        current_input = next_input;
      }
      Err(e) => return Err(e),
    };
  }
  Ok((current_input, result))
}

pub fn parse_gradient_bevel_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::GradientBevel> {
  use nom::combinator::map;

  let (input, color_count) = map(parse_u8, |x| x as usize)(input)?;
  let (input, gradient) = parse_filter_gradient(input, color_count)?;
  let (input, blur_x) = parse_le_fixed16_p16(input)?;
  let (input, blur_y) = parse_le_fixed16_p16(input)?;
  let (input, angle) = parse_le_fixed16_p16(input)?;
  let (input, distance) = parse_le_fixed16_p16(input)?;
  let (input, strength) = parse_le_fixed8_p8(input)?;

  let (input, flags) = parse_u8(input)?;
  let passes = flags & ((1 << 4) - 1);
  let on_top = (flags & (1 << 4)) != 0;
  let composite_source = (flags & (1 << 5)) != 0;
  let knockout = (flags & (1 << 6)) != 0;
  let inner = (flags & (1 << 7)) != 0;

  Ok((
    input,
    swf::filters::GradientBevel {
      gradient,
      blur_x,
      blur_y,
      angle,
      distance,
      strength,
      inner,
      knockout,
      composite_source,
      on_top,
      passes,
    },
  ))
}

pub fn parse_gradient_glow_filter(input: &[u8]) -> NomResult<&[u8], swf::filters::GradientGlow> {
  use nom::combinator::map;

  let (input, color_count) = map(parse_u8, |x| x as usize)(input)?;
  let (input, gradient) = parse_filter_gradient(input, color_count)?;
  let (input, blur_x) = parse_le_fixed16_p16(input)?;
  let (input, blur_y) = parse_le_fixed16_p16(input)?;
  let (input, angle) = parse_le_fixed16_p16(input)?;
  let (input, distance) = parse_le_fixed16_p16(input)?;
  let (input, strength) = parse_le_fixed8_p8(input)?;

  let (input, flags) = parse_u8(input)?;
  let passes = flags & ((1 << 4) - 1);
  let on_top = (flags & (1 << 4)) != 0;
  let composite_source = (flags & (1 << 5)) != 0;
  let knockout = (flags & (1 << 6)) != 0;
  let inner = (flags & (1 << 7)) != 0;

  Ok((
    input,
    swf::filters::GradientGlow {
      gradient,
      blur_x,
      blur_y,
      angle,
      distance,
      strength,
      inner,
      knockout,
      composite_source,
      on_top,
      passes,
    },
  ))
}
